lib/gitsh/git_command_list.rb

```diff
@@ -0,0 +1,56 @@
+module Gitsh
+  class GitCommandList
+    def initialize(env)
+      @env = env
+    end
+
+    def to_a
+      try_using(commands_from_list_cmds) do
+        try_using(commands_from_help('help -a --no-verbose')) do
+          try_using(commands_from_help('help -a'))
+        end
+      end
+    end
+
+    private
+
+    attr_accessor :env
+
+    def try_using(result, default = [])
+      if result && result.any?
+        result
+      elsif block_given?
+        yield
+      else
+        default
+      end
+    end
+
+    def commands_from_list_cmds
+      git_output('--list-cmds=main,nohelpers').sort
+    end
+
+    def commands_from_help(command)
+      git_output(command).
+        select { |line| line =~ /^  [a-z]/ }.
+        map { |line| line.split(/\s+/) }.
+        flatten.
```

Comment:
I tried re-writing this whole block as one `#inject` -- so one iteration through the elements -- but this `#flatten` defeated me. What you have is much more readable than the `#inject` variant.

Reply:
Thanks for trying the `#inject` route. I'll merge this as-is.